shared-bindings/nvm/ByteArray.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/nvm/ByteArray.h"
+
+//| .. currentmodule:: nvm
+//|
+//| :class:`ByteArray` -- Presents a stretch of non-volatile memory as a bytearray.
+//| ================================================================================
+//|
+//| Non-volatile memory is avialble as a byte array that persists over reloads
+//| and power cycles.
+//|
+//| Usage::
+//|
+//|    import microcontroller
+//|    microcontroller.nvm[0] = 0xcc
+//|
+
+//| .. class:: ByteArray()
+//|
+//|   Not currently dynamically supported. Access one through `microcontroller.nvm`.
+//|
+STATIC mp_obj_t nvm_bytearray_make_new(const mp_obj_type_t *type,
+        mp_uint_t n_args, mp_uint_t n_kw, const mp_obj_t *args) {
+    return mp_const_none;
+}
+
+//|   .. method:: __len__()
+//|
+//|     Return the length. This is used by (`len`)
+//|
+STATIC mp_obj_t nvm_bytearray___len__(mp_obj_t self_in) {
```

Comment:
This should work, but `__len__()` doesn't seem to work right in Circuit/MicroPython:
```
>>> len(microcontroller.nvm)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'ByteArray' has no len()

>>> len(bytearray(7))
7

>>> microcontroller.nvm.__len__()
256

>>> bytearray(7).__len__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'bytearray' object has no attribute '__len__'
>>> [1,2].__len__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute '__len__'

```
It doesn't work in the `unix` build either, so it's not some idiosyncrasy of `atmel-samd`

Reply:
Fixed. Its unary_op internally :-/